karabo_data/geometry2.py

```diff
@@ -0,0 +1,483 @@
+          (lengths 16, 512, 128). ss/fs are slow-scan and fast-scan.
+        """
+        from matplotlib.cm import viridis
+        from matplotlib.backends.backend_agg import FigureCanvasAgg
+        from matplotlib.figure import Figure
+
+        fig = Figure((10, 10))
+        FigureCanvasAgg(fig)
+        ax = fig.add_subplot(1, 1, 1)
+        my_viridis = copy(viridis)
+        # Use a dark grey for missing data
+        my_viridis.set_bad('0.25', 1.)
+
+        res, centre = self.position_all_modules(modules_data)
+        ax.imshow(res, origin='lower', cmap=my_viridis)
+
+        cx, cy = centre
+        ax.hlines(cy, cx - 20, cx + 20, colors='w', linewidths=1)
+        ax.vlines(cx, cy - 20, cy + 20, colors='w', linewidths=1)
+        return fig
+
+    def snap(self):
+        """Snap geometry to a 2D pixel grid
+
+        This returns a new geometry object. The 'snapped' geometry is
+        less accurate, but can assemble data into a 2D array more efficiently,
+        because it doesn't do any interpolation.
+        """
+        new_modules = []
+        for module in self.modules:
+            new_tiles = [t.snap() for t in module]
+            new_modules.append(new_tiles)
+        return AGIPD_1M_SnappedGeometry(new_modules)
+
+class AGIPD_1M_SnappedGeometry:
+    """AGIPD geometry approximated to align modules to a 2D grid
+
+    The coordinates used in this class are (y, x) suitable for indexing a
+    Numpy array; this does not match the (x, y, z) coordinates in the more
+    precise geometry above.
+    """
+    def __init__(self, modules):
+        self.modules = modules
+
+    def position_all_modules(self, data):
+        """Assemble data from this detector according to where the pixels are.
+
+        Parameters
+        ----------
+
+        data : ndarray
+          The last three dimensions should be channelno, pixel_ss, pixel_fs
+          (lengths 16, 512, 128). ss/fs are slow-scan and fast-scan.
+
+        Returns
+        -------
+        out : ndarray
+          Array with one dimension fewer than the input.
+          The last two dimensions represent pixel y and x in the detector space.
+        centre : ndarray
+          (y, x) pixel location of the detector centre in this geometry.
+        """
+        assert data.shape[-3:] == (16, 512, 128)
+        size_yx, centre = self._plotting_dimensions()
+        out = np.full(data.shape[:-3] + size_yx, np.nan, dtype=data.dtype)
+        for i, module in enumerate(self.modules):
+            mod_data = data[..., i, :, :]
+            tiles_data = np.split(mod_data, 8, axis=-2)
+            for j, tile in enumerate(module):
+                tile_data = tiles_data[j]
+                # Offset by centre to make all coordinates positive
+                y, x = tile.corner_idx + centre
+                h, w = tile.pixel_dims
+                out[..., y:y+h, x:x+w] = tile.transform(tile_data)
+
+        return out, centre
+
+    def _plotting_dimensions(self):
```

Comment:
I know it might sound silly, and it doesn't really matter, but I think a method name of something like get_dimensions is better than plotting_dimensions, since it's not only about plotting anymore.